feat: Client-side Proxy wrapper

gselzer opened the pull request:
I have found that *as-is* the `MMCorePlusProxy` object has two shortcomings:
1. It is not (as written) a `CMMCorePlus`. Therefore functionality that does any `isinstance` checking [example](https://github.com/pymmcore-plus/pymmcore-widgets/blob/bbacae1cd6e204b898faa318fac5cc8ba8614b4a/src/pymmcore_widgets/control/_q_stage_controller.py#L88-L91) will fail It's possible that we could add `CMMCorePlus` as a parent of `MMCorePlusProxy`, didn't really look into it, but even if we could:
2. It cannot be used by multiple threads. Naively this might be considered a good thing, however this makes responding to core events tricky and hinders use from GUIs. For example, over in https://github.com/gselzer/pymmcore-remote-sandbox I have a GUI containing an `ImagePreview` and a `StageWidget` attached to a server:
   * Using that `MMCorePlusProxy` in response to GUI events requires its transfer from the main python thread to the Qt thread responding to user input. 
   * When calling a method on the `MMCorePlusProxy` causes a core event to trigger (in my use case, it would be the `imageSnapped` event), a Pyro worker thread would then invoke the callback. Transferring the Proxy to the Pyro thread is a bit sketchy, especially since the Qt thread may still have a handle on it if it is waiting for its called method to return.

At a high level, this change:
* Introduces a base class `ProxyHandler` used to encapsulate multiple `Pyro5.api.Proxy` objects. Instances of this base class can be used by multiple threads *without deadlock*, as under the hood this object redirects all calls to a proxy owned by the calling thread.
* Introduces `ClientCMMCorePlus` and `ClientMDARunner` as implementations of `ProxyHandler`, for interacting with a remote `CMMCorePlus` and `MDARunner`, respectively. Any future pymmcore-remote classes that offer event callbacks should also become implementations of `ProxyHandler.
* Privatizes `MMCorePlusProxy`, as it:
  * No longer manages its own `MDARunnerProxy` (`ClientCMMCorePlus` has its own MDA field to avoid MDA event deadlock) - could consider adding this back later if necessary
  * Suffers from deadlock

gselzer committed:
feat: Client-side Proxy wrapper

gselzer commented:
> next, I'm probably against the general pattern of
> 
> ```python
> class FakeSub(SuperCls):
>     def __getattribute__(self, name: str) -> Any:
>         # redirect elsewhere
> ```

> _just_ to be able to satisfy an isinstance check.

Well that's not the *main* reason I'm doing it, but yeah, definitely ugly, and if it *were* the only reason I don't think this would be a good solution. I'd rather try e.g. mucking `__instancecheck__` (though I did wonder whether Pyro would support throwing `CMMCorePlus` as a supertype of `MMCorePlusProxy`...)

> > It cannot be used by multiple threads.
> 
> it kinda sounds like you're also describing #4?

Oh yeah, look at that, guess I am 😅 

You wrote over there:

> then the only solution I can currently think of is some rather dark magic of mucking with the `locals()` and closures of the callback (swapping in a new proxy object) before calling the callback... and that would require deep mucking with psygnal itself, and is probably incompatible with Qt.

I'd say that this is a lighter version of magic than what you describe. Short of switching the underlying messaging technology (which, again, I'm open to, but I wouldn't be surprised if they have the same limitations), I don't really know how else we might fix this...

tlambert03 commented:
i wouldn't be switching tech to circumvent that issue, it would definitely still have that issue. Pyro is just a deep call-stack with a relatively hard to read codebase, and I'd rather have lower-level direct control when trying to deal with these tricky concurrency issues

tlambert03 commented:
> I'd say that this is a lighter version of magic than what you describe. Short of switching the underlying messaging technology (which, again, I'm open to, but I wouldn't be surprised if they have the same limitations), I don't really know how else we might fix this...

can you break apart the goals of `isinstance` and your proposed fix to #4, so that it's not a literal subclass?

gselzer commented:
> can you break apart the goals of `isinstance` and your proposed fix to #4, so that it's not a literal subclass?

Not really, since I'd be giving you more or less the same thing (I suppose the `CMMCorePlus` supertype of `ClientCMMCorePlus` would go away...). All of the methods on `ClientCMMCorePlus` are necessary for solving #4 - `__getattribute__` and `_call_proxy` (which could be combined if necessary) delegate method calls to an appropriate proxy, and `instance` for Object deserialization.

tlambert03 commented:
> I suppose the CMMCorePlus supertype of ClientCMMCorePlus would go away...

that's specifically what I mean, I'm asking if we can decouple those goals.  can you solve the general issue of threading and callbacks independently?  I'll also note that it's probably not *just* the CMMCorePlus class for which this needs to be solved... also the MDARunner (and anything that generically has events), right?  So breaking it out from the whole subclassing thing makes sense anyway

gselzer committed:
Merge remote-tracking branch 'upstream/main' into proxy-wrapper

gselzer committed:
Add types-cachetools to mypy deps

gselzer committed:
Generalize ProxyHandler

gselzer commented:
> that's specifically what I mean, I'm asking if we can decouple those goals. can you solve the general issue of threading and callbacks independently?

Gotcha, I took the liberty of abstracting the proxy handler into a new class, with the `ClientCMMCorePlus` object becoming a thin subtype. 

> I'll also note that it's probably not _just_ the CMMCorePlus class for which this needs to be solved... also the MDARunner (and anything that generically has events), right? So breaking it out from the whole subclassing thing makes sense anyway

I haven't really looked too much at the `MDARunner` proxy, can look at that soon (looks like the example is broken, will have to fix that too)

gselzer committed:
Misc fixes

Centered around using ClientCMMCorePlus in examples

gselzer committed:
Use ClientCMMCorePlus in tests

gselzer committed:
Put default URI in ClientCMMCorePlus

gselzer committed:
Test calling core in core callback

This test was adapted from https://github.com/pymmcore-plus/pymmcore-remote/issues/4

gselzer committed:
Merge remote-tracking branch 'upstream/main' into proxy-wrapper

gselzer committed:
Clean up remote_mda example

Don't need the URI anymore

gselzer committed:
Update lockfile

gselzer committed:
Actually, remove lockfile

Maybe address lockfile in another PR

gselzer committed:
Implement MDARunner proxy handler

And test that it works

gselzer commented:
@tlambert03 let me know what you think about the new `ProxyHandler` class.

I'm confident that this closes #4 - added the problematic test that arose in the issue.

> I'll also note that it's probably not _just_ the CMMCorePlus class for which this needs to be solved... also the MDARunner (and anything that generically has events), right? So breaking it out from the whole subclassing thing makes sense anyway

Yup, hadn't thought about it because I hadn't had the need for that yet, but added a `ClientMDARunner` as well.

gselzer committed:
Remove MMCorerPlusProxy from the public attrs

Users should be using ClientCMMCorePlus

gselzer committed:
Add comment about _handler_fields

gselzer committed:
Rename call_proxy to _proxy_attr

Better fitting name. Also added docstring for clarity

gselzer committed:
Remove _handler_fields

We can just use __getattr__, duh

gselzer committed:
Clarify context manager methods

gselzer commented on src/pymmcore_remote/client.py:
```diff
@@ -20,11 +23,11 @@
         events = ClientSideMDASignaler()
         object.__setattr__(self, "events", events)
-        cb_thread.api_daemon.register(events)
+        _DaemonThread.instance("CallbackDaemon").api_daemon.register(events)
```
Assuming it is fine for:
1. All MDA and MMCore proxies to share one thread
2. For the events signaler to be registered  after starting the thread

Can reevaluate if either of those assumptions are no good.

gselzer committed:
Remove _pyroClaimOwnership override

The field we needed to herd no longer exists

tlambert03 reviewed (approved):
i'm grateful to have any attention at all on this repo, so I'm inclined to merge and let you run with this for a bit.  I can't say I understand everything that's going on here, but probably won't have time to give it proper attention soon.

Can you just update the top post on this issue, summarizing in a couple short bullet points what changes you've made here at a high level and why? and then we can merge.

also: is the changing of the main public API (MMCorePlusProxy -> ClientCMMCorePlus) indicative of a deep restructuring?  or more of a preference?   Does the readme need to be updated?

gselzer commented:
> i'm grateful to have any attention at all on this repo, so I'm inclined to merge and let you run with this for a bit. I can't say I understand everything that's going on here, but probably won't have time to give it proper attention soon.

Appreciate it. This changeset may not be the ideal solution, but it is definitely a step forward, and it enables our use cases, so I'd be happy to merge. And I would greatly appreciate any attention you can provide at a later date.

> Can you just update the top post on this issue, summarizing in a couple short bullet points what changes you've made here at a high level and why? and then we can merge.

Yeah, will do

> also: is the changing of the main public API (MMCorePlusProxy -> ClientCMMCorePlus) indicative of a deep restructuring? or more of a preference? Does the readme need to be updated?

Well, the goal is to direct users towards objects that are not vulnerable to deadlock. Do you see any issues with a straight removal from the public API at this point? And yeah, we do need to update the `README` if so...

tlambert03 commented:
> Well, the goal is to direct users towards objects that are not vulnerable to deadlock. Do you see any issues with a straight removal from the public API at this point?

at this point, I wouldn't want the end user to know about more than one client object.  and it should "behave" exactly like a CMMCorePlus object... that's really the only requirement, and whatever tricks are needed under the hood are fine with me.  As long as that end-user API stays constant, go nuts :)

you can merge when ready!

gselzer committed:
Make MMCorePlusProxy, MDARunnerProxy private

Don't want users to use them because without management they suffer from
deadlock

gselzer committed:
Update README - use ClientCMMCorePlus
